File: Shell_Sort/Ruby/majjoha/shell_sort.rb
def shell_sort(array)
  h = array.size / 2
  while h > 0 do
    (h..array.length).each_with_index do |item, j|
      val = array[j]
      i = j
      while (i >= h && array[i-h] > val)
        array[i] = array[i-h]
        i -= h
      end
      array[i] = val
    end
    h /= 2
  end
  array
end

File: Shell_Sort/Go/jcla1/shell_sort.go
package shellSort

func ShellSort(arr []int) []int {
	h := len(arr) / 2
	var val, i int

	for h > 0 {
		for j := h; j < len(arr); j++ {
			val = arr[j]
			i = j

			for i >= h && arr[i-h] > val {
				arr[i] = arr[i-h]
				i -= h
			}

			arr[i] = val
		}

		h /= 2
	}
	return arr
}
